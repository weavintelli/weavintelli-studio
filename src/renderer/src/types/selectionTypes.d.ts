export type TriggerMode = 'selected' | 'ctrlkey' | 'shortcut'
export type FilterMode = 'default' | 'whitelist' | 'blacklist'
export interface ActionItem {
  id: string
  name: string
  enabled: boolean
  isBuiltIn: boolean
  icon?: string
  prompt?: string
  assistantId?: string
  selectedText?: string
  searchEngine?: string
}

export interface SelectionState {
  selectionEnabled: boolean
  triggerMode: TriggerMode
  isCompact: boolean
  isAutoClose: boolean
  isAutoPin: boolean
  isFollowToolbar: boolean
  isRemeberWinSize: boolean
  filterMode: FilterMode
  filterList: string[]
  actionWindowOpacity: number
  actionItems: ActionItem[]
}
