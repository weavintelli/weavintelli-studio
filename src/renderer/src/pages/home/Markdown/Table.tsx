import store from '@renderer/store'
import { messageBlocksSelectors } from '@renderer/store/messageBlock'
import { Tooltip } from 'antd'
import { Check, Copy } from 'lucide-react'
import React, { memo, useCallback, useState } from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'

interface Props {
  children: React.ReactNode
  node?: any
  blockId?: string
}

/**
 * 自定义 Markdown 表格组件，提供 copy 功能。
 */
const Table: React.FC<Props> = ({ children, node, blockId }) => {
  const { t } = useTranslation()
  const [copied, setCopied] = useState(false)

  const handleCopyTable = useCallback(() => {
    const tableMarkdown = extractTableMarkdown(blockId ?? '', node?.position)
    if (!tableMarkdown) return

    navigator.clipboard
      .writeText(tableMarkdown)
      .then(() => {
        setCopied(true)
        setTimeout(() => setCopied(false), 2000)
      })
      .catch((error) => {
        window.message?.error({ content: `${t('message.copy.failed')}: ${error}`, key: 'copy-table-error' })
      })
  }, [node, blockId, t])

  return (
    <TableWrapper className="table-wrapper">
      <table>{children}</table>
      <ToolbarWrapper className="table-toolbar">
        <Tooltip title={t('common.copy')} mouseEnterDelay={0.8}>
          <ToolButton role="button" aria-label={t('common.copy')} onClick={handleCopyTable}>
            {copied ? (
              <Check size={14} style={{ color: 'var(--color-primary)' }} data-testid="check-icon" />
            ) : (
              <Copy size={14} data-testid="copy-icon" />
            )}
          </ToolButton>
        </Tooltip>
      </ToolbarWrapper>
    </TableWrapper>
  )
}

/**
 * 从原始 Markdown 内容中提取表格源代码
 * @param blockId 消息块 ID
 * @param position 表格节点的位置信息
 * @returns 源代码
 */
export function extractTableMarkdown(blockId: string, position: any): string {
  if (!position || !blockId) return ''

  const block = messageBlocksSelectors.selectById(store.getState(), blockId)

  if (!block || !('content' in block) || typeof block.content !== 'string') return ''

  const { start, end } = position
  const lines = block.content.split('\n')

  // 提取表格对应的行（行号从1开始，数组索引从0开始）
  const tableLines = lines.slice(start.line - 1, end.line)
  return tableLines.join('\n').trim()
}

const TableWrapper = styled.div`
  position: relative;

  .table-toolbar {
    border-radius: 4px;
    opacity: 0;
    transition: opacity 0.2s ease;
    transform: translateZ(0);
    will-change: opacity;
  }
  &:hover {
    .table-toolbar {
      opacity: 1;
    }
  }
`

const ToolbarWrapper = styled.div`
  position: absolute;
  top: 8px;
  right: 8px;
  z-index: 10;
`

const ToolButton = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 4px;
  cursor: pointer;
  user-select: none;
  transition: all 0.2s ease;
  opacity: 1;
  color: var(--color-text-3);
  background-color: var(--color-background-mute);
  will-change: background-color, opacity;

  &:hover {
    background-color: var(--color-background-soft);
  }
`

export default memo(Table)
