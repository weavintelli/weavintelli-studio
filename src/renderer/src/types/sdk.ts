import Anthropic from '@anthropic-ai/sdk'
import {
  Message,
  MessageCreateParams,
  MessageParam,
  RawMessageStreamEvent,
  ToolUnion,
  ToolUseBlock
} from '@anthropic-ai/sdk/resources'
import { MessageStream } from '@anthropic-ai/sdk/resources/messages/messages'
import {
  Content,
  CreateChatParameters,
  FunctionCall,
  GenerateContentResponse,
  GoogleGenAI,
  Model as GeminiModel,
  SendMessageParameters,
  Tool
} from '@google/genai'
import OpenAI, { AzureOpenAI } from 'openai'
import { Stream } from 'openai/streaming'

export type SdkInstance = OpenAI | AzureOpenAI | Anthropic | GoogleGenAI
export type SdkParams = OpenAISdkParams | OpenAIResponseSdkParams | AnthropicSdkParams | GeminiSdkParams
export type SdkRawChunk = OpenAISdkRawChunk | OpenAIResponseSdkRawChunk | AnthropicSdkRawChunk | GeminiSdkRawChunk
export type SdkRawOutput = OpenAISdkRawOutput | OpenAIResponseSdkRawOutput | AnthropicSdkRawOutput | GeminiSdkRawOutput
export type SdkMessageParam =
  | OpenAISdkMessageParam
  | OpenAIResponseSdkMessageParam
  | AnthropicSdkMessageParam
  | GeminiSdkMessageParam
export type SdkToolCall =
  | OpenAI.Chat.Completions.ChatCompletionMessageToolCall
  | ToolUseBlock
  | FunctionCall
  | OpenAIResponseSdkToolCall
export type SdkTool = OpenAI.Chat.Completions.ChatCompletionTool | ToolUnion | Tool | OpenAIResponseSdkTool
export type SdkModel = OpenAI.Models.Model | Anthropic.ModelInfo | GeminiModel

export type RequestOptions = Anthropic.RequestOptions | OpenAI.RequestOptions | GeminiOptions

/**
 * OpenAI
 */

type OpenAIParamsWithoutReasoningEffort = Omit<OpenAI.Chat.Completions.ChatCompletionCreateParams, 'reasoning_effort'>

export type ReasoningEffortOptionalParams = {
  thinking?: { type: 'disabled' | 'enabled' | 'auto'; budget_tokens?: number }
  reasoning?: { max_tokens?: number; exclude?: boolean; effort?: string } | OpenAI.Reasoning
  reasoning_effort?: OpenAI.Chat.Completions.ChatCompletionCreateParams['reasoning_effort'] | 'none' | 'auto'
  enable_thinking?: boolean
  thinking_budget?: number
  enable_reasoning?: boolean
  // Add any other potential reasoning-related keys here if they exist
}

export type OpenAISdkParams = OpenAIParamsWithoutReasoningEffort & ReasoningEffortOptionalParams
export type OpenAISdkRawChunk =
  | OpenAI.Chat.Completions.ChatCompletionChunk
  | ({
      _request_id?: string | null | undefined
    } & OpenAI.ChatCompletion)

export type OpenAISdkRawOutput = Stream<OpenAI.Chat.Completions.ChatCompletionChunk> | OpenAI.ChatCompletion
export type OpenAISdkRawContentSource =
  | OpenAI.Chat.Completions.ChatCompletionChunk.Choice.Delta
  | OpenAI.Chat.Completions.ChatCompletionMessage

export type OpenAISdkMessageParam = OpenAI.Chat.Completions.ChatCompletionMessageParam

/**
 * OpenAI Response
 */

export type OpenAIResponseSdkParams = OpenAI.Responses.ResponseCreateParams
export type OpenAIResponseSdkRawOutput = Stream<OpenAI.Responses.ResponseStreamEvent> | OpenAI.Responses.Response
export type OpenAIResponseSdkRawChunk = OpenAI.Responses.ResponseStreamEvent | OpenAI.Responses.Response
export type OpenAIResponseSdkMessageParam = OpenAI.Responses.ResponseInputItem
export type OpenAIResponseSdkToolCall = OpenAI.Responses.ResponseFunctionToolCall
export type OpenAIResponseSdkTool = OpenAI.Responses.Tool

/**
 * Anthropic
 */

export type AnthropicSdkParams = MessageCreateParams
export type AnthropicSdkRawOutput = MessageStream | Message
export type AnthropicSdkRawChunk = RawMessageStreamEvent | Message
export type AnthropicSdkMessageParam = MessageParam

/**
 * Gemini
 */

export type GeminiSdkParams = SendMessageParameters & CreateChatParameters
export type GeminiSdkRawOutput = AsyncGenerator<GenerateContentResponse> | GenerateContentResponse
export type GeminiSdkRawChunk = GenerateContentResponse
export type GeminiSdkMessageParam = Content
export type GeminiSdkToolCall = FunctionCall

export type GeminiOptions = {
  streamOutput: boolean
  abortSignal?: AbortSignal
  timeout?: number
}
