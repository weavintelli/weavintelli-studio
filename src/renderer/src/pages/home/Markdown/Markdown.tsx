import 'katex/dist/katex.min.css'
import 'katex/dist/contrib/copy-tex'
import 'katex/dist/contrib/mhchem'

import ImageViewer from '@renderer/components/ImageViewer'
import MarkdownShadowDOMRenderer from '@renderer/components/MarkdownShadowDOMRenderer'
import { useSettings } from '@renderer/hooks/useSettings'
import { EVENT_NAMES, EventEmitter } from '@renderer/services/EventService'
import type { MainTextMessageBlock, ThinkingMessageBlock, TranslationMessageBlock } from '@renderer/types/newMessage'
import { parseJSON } from '@renderer/utils'
import { escapeBrackets, removeSvgEmptyLines } from '@renderer/utils/formats'
import { findCitationInChildren, getCodeBlockId } from '@renderer/utils/markdown'
import { isEmpty } from 'lodash'
import { type FC, memo, useCallback, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import ReactMarkdown, { type Components, defaultUrlTransform } from 'react-markdown'
import rehypeKatex from 'rehype-katex'
// @ts-ignore rehype-mathjax is not typed
import rehypeMathjax from 'rehype-mathjax'
import rehypeRaw from 'rehype-raw'
import remarkCjkFriendly from 'remark-cjk-friendly'
import remarkGfm from 'remark-gfm'
import remarkMath from 'remark-math'

import CodeBlock from './CodeBlock'
import Link from './Link'
import Table from './Table'

const ALLOWED_ELEMENTS =
  /<(style|p|div|span|b|i|strong|em|ul|ol|li|table|tr|td|th|thead|tbody|h[1-6]|blockquote|pre|code|br|hr|svg|path|circle|rect|line|polyline|polygon|text|g|defs|title|desc|tspan|sub|sup)/i
const DISALLOWED_ELEMENTS = ['iframe']

interface Props {
  // message: Message & { content: string }
  block: MainTextMessageBlock | TranslationMessageBlock | ThinkingMessageBlock
}

const Markdown: FC<Props> = ({ block }) => {
  const { t } = useTranslation()
  const { mathEngine } = useSettings()

  const remarkPlugins = useMemo(() => {
    const plugins = [remarkGfm, remarkCjkFriendly]
    if (mathEngine !== 'none') {
      plugins.push(remarkMath)
    }
    return plugins
  }, [mathEngine])

  const messageContent = useMemo(() => {
    const empty = isEmpty(block.content)
    const paused = block.status === 'paused'
    const content = empty && paused ? t('message.chat.completion.paused') : block.content
    return removeSvgEmptyLines(escapeBrackets(content))
  }, [block, t])

  const rehypePlugins = useMemo(() => {
    const plugins: any[] = []
    if (ALLOWED_ELEMENTS.test(messageContent)) {
      plugins.push(rehypeRaw)
    }
    if (mathEngine === 'KaTeX') {
      plugins.push(rehypeKatex as any)
    } else if (mathEngine === 'MathJax') {
      plugins.push(rehypeMathjax as any)
    }
    return plugins
  }, [mathEngine, messageContent])

  const onSaveCodeBlock = useCallback(
    (id: string, newContent: string) => {
      EventEmitter.emit(EVENT_NAMES.EDIT_CODE_BLOCK, {
        msgBlockId: block.id,
        codeBlockId: id,
        newContent
      })
    },
    [block.id]
  )

  const components = useMemo(() => {
    return {
      a: (props: any) => <Link {...props} citationData={parseJSON(findCitationInChildren(props.children))} />,
      code: (props: any) => (
        <CodeBlock {...props} id={getCodeBlockId(props?.node?.position?.start)} onSave={onSaveCodeBlock} />
      ),
      table: (props: any) => <Table {...props} blockId={block.id} />,
      img: (props: any) => <ImageViewer style={{ maxWidth: 500, maxHeight: 500 }} {...props} />,
      pre: (props: any) => <pre style={{ overflow: 'visible' }} {...props} />,
      p: (props) => {
        const hasImage = props?.node?.children?.some((child: any) => child.tagName === 'img')
        if (hasImage) return <div {...props} />
        return <p {...props} />
      }
    } as Partial<Components>
  }, [onSaveCodeBlock, block.id])

  if (messageContent.includes('<style>')) {
    components.style = MarkdownShadowDOMRenderer as any
  }

  const urlTransform = useCallback((value: string) => {
    if (value.startsWith('data:image/png') || value.startsWith('data:image/jpeg')) return value
    return defaultUrlTransform(value)
  }, [])

  return (
    <ReactMarkdown
      rehypePlugins={rehypePlugins}
      remarkPlugins={remarkPlugins}
      className="markdown"
      components={components}
      disallowedElements={DISALLOWED_ELEMENTS}
      urlTransform={urlTransform}
      remarkRehypeOptions={{
        footnoteLabel: t('common.footnotes'),
        footnoteLabelTagName: 'h4',
        footnoteBackContent: ' '
      }}>
      {messageContent}
    </ReactMarkdown>
  )
}

export default memo(Markdown)
