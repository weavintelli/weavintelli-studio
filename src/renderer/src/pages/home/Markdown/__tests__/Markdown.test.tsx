import 'katex/dist/katex.min.css'

import type { MainTextMessageBlock, ThinkingMessageBlock, TranslationMessageBlock } from '@renderer/types/newMessage'
import { MessageBlockStatus, MessageBlockType } from '@renderer/types/newMessage'
import { render, screen } from '@testing-library/react'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

import Markdown from '../Markdown'

// Mock dependencies
const mockUseSettings = vi.fn()
const mockUseTranslation = vi.fn()

// Mock hooks
vi.mock('@renderer/hooks/useSettings', () => ({
  useSettings: () => mockUseSettings()
}))

vi.mock('react-i18next', () => ({
  useTranslation: () => mockUseTranslation()
}))

// Mock services
vi.mock('@renderer/services/EventService', () => ({
  EVENT_NAMES: {
    EDIT_CODE_BLOCK: 'EDIT_CODE_BLOCK'
  },
  EventEmitter: {
    emit: vi.fn()
  }
}))

// Mock utilities
vi.mock('@renderer/utils', () => ({
  parseJSON: vi.fn((str) => {
    try {
      return JSON.parse(str || '{}')
    } catch {
      return {}
    }
  })
}))

vi.mock('@renderer/utils/formats', () => ({
  escapeBrackets: vi.fn((str) => str),
  removeSvgEmptyLines: vi.fn((str) => str)
}))

vi.mock('@renderer/utils/markdown', () => ({
  findCitationInChildren: vi.fn(() => '{"id": 1, "url": "https://example.com"}'),
  getCodeBlockId: vi.fn(() => 'code-block-1')
}))

// Mock components with more realistic behavior
vi.mock('../CodeBlock', () => ({
  __esModule: true,
  default: ({ id, onSave, children }: any) => (
    <div data-testid="code-block" data-id={id}>
      <code>{children}</code>
      <button type="button" onClick={() => onSave(id, 'new content')}>
        Save
      </button>
    </div>
  )
}))

vi.mock('../ImagePreview', () => ({
  __esModule: true,
  default: (props: any) => <img data-testid="image-preview" {...props} />
}))

vi.mock('../Link', () => ({
  __esModule: true,
  default: ({ citationData, children, ...props }: any) => (
    <a data-testid="citation-link" data-citation={citationData} {...props}>
      {children}
    </a>
  )
}))

vi.mock('../Table', () => ({
  __esModule: true,
  default: ({ children, blockId }: any) => (
    <div data-testid="table-component" data-block-id={blockId}>
      <table>{children}</table>
      <button type="button" data-testid="copy-table-button">
        Copy Table
      </button>
    </div>
  )
}))

vi.mock('@renderer/components/MarkdownShadowDOMRenderer', () => ({
  __esModule: true,
  default: ({ children }: any) => <div data-testid="shadow-dom">{children}</div>
}))

// Mock plugins
vi.mock('remark-gfm', () => ({ __esModule: true, default: vi.fn() }))
vi.mock('remark-cjk-friendly', () => ({ __esModule: true, default: vi.fn() }))
vi.mock('remark-math', () => ({ __esModule: true, default: vi.fn() }))
vi.mock('rehype-katex', () => ({ __esModule: true, default: vi.fn() }))
vi.mock('rehype-mathjax', () => ({ __esModule: true, default: vi.fn() }))
vi.mock('rehype-raw', () => ({ __esModule: true, default: vi.fn() }))

// Mock ReactMarkdown with realistic rendering
vi.mock('react-markdown', () => ({
  __esModule: true,
  default: ({ children, components, className }: any) => (
    <div data-testid="markdown-content" className={className}>
      {children}
      {/* Simulate component rendering */}
      {components?.a && <span data-testid="has-link-component">link</span>}
      {components?.code && (
        <div data-testid="has-code-component">
          {components.code({ children: 'test code', node: { position: { start: { line: 1 } } } })}
        </div>
      )}
      {components?.table && (
        <div data-testid="has-table-component">
          {components.table({ children: 'test table', node: { position: { start: { line: 1 } } } })}
        </div>
      )}
      {components?.img && <span data-testid="has-img-component">img</span>}
      {components?.style && <span data-testid="has-style-component">style</span>}
    </div>
  )
}))

describe('Markdown', () => {
  let mockEventEmitter: any

  beforeEach(async () => {
    vi.clearAllMocks()

    // Default settings
    mockUseSettings.mockReturnValue({ mathEngine: 'KaTeX' })
    mockUseTranslation.mockReturnValue({
      t: (key: string) => (key === 'message.chat.completion.paused' ? 'Paused' : key)
    })

    // Get mocked EventEmitter
    const { EventEmitter } = await import('@renderer/services/EventService')
    mockEventEmitter = EventEmitter
  })

  afterEach(() => {
    vi.restoreAllMocks()
  })

  // Test data helpers
  const createMainTextBlock = (overrides: Partial<MainTextMessageBlock> = {}): MainTextMessageBlock => ({
    id: 'test-block-1',
    messageId: 'test-message-1',
    type: MessageBlockType.MAIN_TEXT,
    status: MessageBlockStatus.SUCCESS,
    createdAt: new Date().toISOString(),
    content: '# Test Markdown\n\nThis is **bold** text.',
    ...overrides
  })

  describe('rendering', () => {
    it('should render markdown content with correct structure', () => {
      const block = createMainTextBlock({ content: 'Test content' })
      render(<Markdown block={block} />)

      const markdown = screen.getByTestId('markdown-content')
      expect(markdown).toBeInTheDocument()
      expect(markdown).toHaveClass('markdown')
      expect(markdown).toHaveTextContent('Test content')
    })

    it('should handle empty content gracefully', () => {
      const block = createMainTextBlock({ content: '' })

      expect(() => render(<Markdown block={block} />)).not.toThrow()

      const markdown = screen.getByTestId('markdown-content')
      expect(markdown).toBeInTheDocument()
    })

    it('should show paused message when content is empty and status is paused', () => {
      const block = createMainTextBlock({
        content: '',
        status: MessageBlockStatus.PAUSED
      })
      render(<Markdown block={block} />)

      const markdown = screen.getByTestId('markdown-content')
      expect(markdown).toHaveTextContent('Paused')
    })

    it('should prioritize actual content over paused status', () => {
      const block = createMainTextBlock({
        content: 'Real content',
        status: MessageBlockStatus.PAUSED
      })
      render(<Markdown block={block} />)

      const markdown = screen.getByTestId('markdown-content')
      expect(markdown).toHaveTextContent('Real content')
      expect(markdown).not.toHaveTextContent('Paused')
    })

    it('should process content through format utilities', async () => {
      const { escapeBrackets, removeSvgEmptyLines } = await import('@renderer/utils/formats')
      const content = 'Content with [brackets] and SVG'

      render(<Markdown block={createMainTextBlock({ content })} />)

      expect(escapeBrackets).toHaveBeenCalledWith(content)
      expect(removeSvgEmptyLines).toHaveBeenCalledWith(content)
    })

    it('should match snapshot', () => {
      const { container } = render(<Markdown block={createMainTextBlock()} />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('block type support', () => {
    const testCases = [
      {
        name: 'MainTextMessageBlock',
        block: createMainTextBlock({ content: 'Main text content' }),
        expectedContent: 'Main text content'
      },
      {
        name: 'ThinkingMessageBlock',
        block: {
          id: 'thinking-1',
          messageId: 'msg-1',
          type: MessageBlockType.THINKING,
          status: MessageBlockStatus.SUCCESS,
          createdAt: new Date().toISOString(),
          content: 'Thinking content',
          thinking_millsec: 5000
        } as ThinkingMessageBlock,
        expectedContent: 'Thinking content'
      },
      {
        name: 'TranslationMessageBlock',
        block: {
          id: 'translation-1',
          messageId: 'msg-1',
          type: MessageBlockType.TRANSLATION,
          status: MessageBlockStatus.SUCCESS,
          createdAt: new Date().toISOString(),
          content: 'Translated content',
          targetLanguage: 'en'
        } as TranslationMessageBlock,
        expectedContent: 'Translated content'
      }
    ]

    testCases.forEach(({ name, block, expectedContent }) => {
      it(`should handle ${name} correctly`, () => {
        render(<Markdown block={block} />)

        const markdown = screen.getByTestId('markdown-content')
        expect(markdown).toBeInTheDocument()
        expect(markdown).toHaveTextContent(expectedContent)
      })
    })
  })

  describe('math engine configuration', () => {
    it('should configure KaTeX when mathEngine is KaTeX', () => {
      mockUseSettings.mockReturnValue({ mathEngine: 'KaTeX' })

      render(<Markdown block={createMainTextBlock()} />)

      // Component should render successfully with KaTeX configuration
      expect(screen.getByTestId('markdown-content')).toBeInTheDocument()
    })

    it('should configure MathJax when mathEngine is MathJax', () => {
      mockUseSettings.mockReturnValue({ mathEngine: 'MathJax' })

      render(<Markdown block={createMainTextBlock()} />)

      // Component should render successfully with MathJax configuration
      expect(screen.getByTestId('markdown-content')).toBeInTheDocument()
    })

    it('should not load math plugins when mathEngine is none', () => {
      mockUseSettings.mockReturnValue({ mathEngine: 'none' })

      render(<Markdown block={createMainTextBlock()} />)

      // Component should render successfully without math plugins
      expect(screen.getByTestId('markdown-content')).toBeInTheDocument()
    })
  })

  describe('custom components', () => {
    it('should integrate Link component for citations', () => {
      render(<Markdown block={createMainTextBlock()} />)

      expect(screen.getByTestId('has-link-component')).toBeInTheDocument()
    })

    it('should integrate CodeBlock component with edit functionality', () => {
      const block = createMainTextBlock({ id: 'test-block-123' })
      render(<Markdown block={block} />)

      expect(screen.getByTestId('has-code-component')).toBeInTheDocument()

      // Test code block edit event
      const saveButton = screen.getByText('Save')
      saveButton.click()

      expect(mockEventEmitter.emit).toHaveBeenCalledWith('EDIT_CODE_BLOCK', {
        msgBlockId: 'test-block-123',
        codeBlockId: 'code-block-1',
        newContent: 'new content'
      })
    })

    it('should integrate Table component with copy functionality', () => {
      const block = createMainTextBlock({ id: 'test-block-456' })
      render(<Markdown block={block} />)

      expect(screen.getByTestId('has-table-component')).toBeInTheDocument()

      const tableComponent = screen.getByTestId('table-component')
      expect(tableComponent).toHaveAttribute('data-block-id', 'test-block-456')
    })

    it('should integrate ImagePreview component', () => {
      render(<Markdown block={createMainTextBlock()} />)

      expect(screen.getByTestId('has-img-component')).toBeInTheDocument()
    })

    it('should handle style tags with Shadow DOM', () => {
      const block = createMainTextBlock({ content: '<style>body { color: red; }</style>' })
      render(<Markdown block={block} />)

      expect(screen.getByTestId('has-style-component')).toBeInTheDocument()
    })
  })

  describe('HTML content support', () => {
    it('should handle mixed markdown and HTML content', () => {
      const block = createMainTextBlock({
        content: '# Header\n<div>HTML content</div>\n**Bold text**'
      })

      expect(() => render(<Markdown block={block} />)).not.toThrow()

      const markdown = screen.getByTestId('markdown-content')
      expect(markdown).toBeInTheDocument()
      expect(markdown).toHaveTextContent('# Header')
      expect(markdown).toHaveTextContent('HTML content')
      expect(markdown).toHaveTextContent('**Bold text**')
    })

    it('should handle malformed content gracefully', () => {
      const block = createMainTextBlock({
        content: '<unclosed-tag>content\n# Invalid markdown **unclosed'
      })

      expect(() => render(<Markdown block={block} />)).not.toThrow()

      const markdown = screen.getByTestId('markdown-content')
      expect(markdown).toBeInTheDocument()
    })
  })

  describe('component behavior', () => {
    it('should re-render when content changes', () => {
      const { rerender } = render(<Markdown block={createMainTextBlock({ content: 'Initial' })} />)

      expect(screen.getByTestId('markdown-content')).toHaveTextContent('Initial')

      rerender(<Markdown block={createMainTextBlock({ content: 'Updated' })} />)

      expect(screen.getByTestId('markdown-content')).toHaveTextContent('Updated')
    })

    it('should re-render when math engine changes', () => {
      mockUseSettings.mockReturnValue({ mathEngine: 'KaTeX' })
      const { rerender } = render(<Markdown block={createMainTextBlock()} />)

      expect(screen.getByTestId('markdown-content')).toBeInTheDocument()

      mockUseSettings.mockReturnValue({ mathEngine: 'MathJax' })
      rerender(<Markdown block={createMainTextBlock()} />)

      // Should still render correctly with new math engine
      expect(screen.getByTestId('markdown-content')).toBeInTheDocument()
    })
  })
})
