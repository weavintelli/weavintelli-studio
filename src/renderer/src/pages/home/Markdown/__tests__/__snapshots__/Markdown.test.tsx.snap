// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Markdown > rendering > should match snapshot 1`] = `
<div
  class="markdown"
  data-testid="markdown-content"
>
  # Test Markdown

This is **bold** text.
  <span
    data-testid="has-link-component"
  >
    link
  </span>
  <div
    data-testid="has-code-component"
  >
    <div
      data-id="code-block-1"
      data-testid="code-block"
    >
      <code>
        test code
      </code>
      <button
        type="button"
      >
        Save
      </button>
    </div>
  </div>
  <div
    data-testid="has-table-component"
  >
    <div
      data-block-id="test-block-1"
      data-testid="table-component"
    >
      <table>
        test table
      </table>
      <button
        data-testid="copy-table-button"
        type="button"
      >
        Copy Table
      </button>
    </div>
  </div>
  <span
    data-testid="has-img-component"
  >
    img
  </span>
</div>
`;
